File: apps/web/src/lib/rbac/role-permissions.test.ts
import { describe, it, expect } from "vitest";
import { UserRole } from "@/types/api";
import { Permission } from "./permissions";
import {
  ROLE_PERMISSIONS,
  hasPermission,
  hasAllPermissions,
  hasAnyPermission,
} from "./role-permissions";

describe("ROLE_PERMISSIONS", () => {
  it("should give ADMIN all permissions", () => {
    const adminPermissions = ROLE_PERMISSIONS[UserRole.ADMIN];
    const allPermissions = Object.values(Permission);

    expect(adminPermissions.size).toBe(allPermissions.length);
    allPermissions.forEach((permission) => {
      expect(adminPermissions.has(permission)).toBe(true);
    });
  });

  it("should give EMPLOYEE limited permissions", () => {
    const employeePermissions = ROLE_PERMISSIONS[UserRole.EMPLOYEE];

    // Should have these permissions
    expect(employeePermissions.has(Permission.PRODUCTS_VIEW)).toBe(true);
    expect(employeePermissions.has(Permission.STORAGE_VIEW)).toBe(true);
    expect(employeePermissions.has(Permission.SHIPMENTS_VIEW)).toBe(true);
    expect(employeePermissions.has(Permission.SHIPMENTS_CREATE)).toBe(true);
    expect(employeePermissions.has(Permission.SHIPMENTS_UPDATE)).toBe(true);

    // Should NOT have these permissions
    expect(employeePermissions.has(Permission.DASHBOARD_VIEW)).toBe(false);
    expect(employeePermissions.has(Permission.PRODUCTS_CREATE)).toBe(false);
    expect(employeePermissions.has(Permission.ANALYTICS_VIEW)).toBe(false);
    expect(employeePermissions.has(Permission.TEAM_VIEW)).toBe(false);
    expect(employeePermissions.has(Permission.SETTINGS_VIEW)).toBe(false);
  });
});

describe("hasPermission", () => {
  it("should return true when role has the permission", () => {
    expect(hasPermission(UserRole.ADMIN, Permission.DASHBOARD_VIEW)).toBe(true);
    expect(hasPermission(UserRole.EMPLOYEE, Permission.PRODUCTS_VIEW)).toBe(
      true
    );
  });

  it("should return false when role lacks the permission", () => {
    expect(hasPermission(UserRole.EMPLOYEE, Permission.DASHBOARD_VIEW)).toBe(
      false
    );
    expect(hasPermission(UserRole.EMPLOYEE, Permission.PRODUCTS_CREATE)).toBe(
      false
    );
  });

  it("should return false when role is undefined", () => {
    expect(hasPermission(undefined, Permission.DASHBOARD_VIEW)).toBe(false);
  });
});

describe("hasAllPermissions", () => {
  it("should return true when role has all permissions", () => {
    expect(
      hasAllPermissions(UserRole.ADMIN, [
        Permission.DASHBOARD_VIEW,
        Permission.ANALYTICS_VIEW,
      ])
    ).toBe(true);

    expect(
      hasAllPermissions(UserRole.EMPLOYEE, [
        Permission.PRODUCTS_VIEW,
        Permission.STORAGE_VIEW,
      ])
    ).toBe(true);
  });

  it("should return false when role lacks any permission", () => {
    expect(
      hasAllPermissions(UserRole.EMPLOYEE, [
        Permission.PRODUCTS_VIEW,
        Permission.PRODUCTS_CREATE, // Employee doesn't have this
      ])
    ).toBe(false);
  });

  it("should return true for empty permissions array", () => {
    expect(hasAllPermissions(UserRole.EMPLOYEE, [])).toBe(true);
  });

  it("should return false when role is undefined", () => {
    expect(hasAllPermissions(undefined, [Permission.PRODUCTS_VIEW])).toBe(
      false
    );
  });
});

describe("hasAnyPermission", () => {
  it("should return true when role has at least one permission", () => {
    expect(
      hasAnyPermission(UserRole.EMPLOYEE, [
        Permission.PRODUCTS_CREATE, // Doesn't have
        Permission.PRODUCTS_VIEW, // Has this
      ])
    ).toBe(true);
  });

  it("should return false when role has none of the permissions", () => {
    expect(
      hasAnyPermission(UserRole.EMPLOYEE, [
        Permission.DASHBOARD_VIEW,
        Permission.ANALYTICS_VIEW,
      ])
    ).toBe(false);
  });

  it("should return false for empty permissions array", () => {
    expect(hasAnyPermission(UserRole.EMPLOYEE, [])).toBe(false);
  });

  it("should return false when role is undefined", () => {
    expect(hasAnyPermission(undefined, [Permission.PRODUCTS_VIEW])).toBe(false);
  });
});

File: services/inventory-service/src/test/java/com/mirai/inventoryservice/identity/domain/RolePermissionsTest.java
package com.mirai.inventoryservice.identity.domain;

import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Cross-side parity assertions ported from apps/web/src/lib/rbac/role-permissions.test.ts.
 * Both definitions must stay identical - if a permission changes on one side without the
 * other, one of these two suites should fail.
 */
class RolePermissionsTest {

    @Test
    void adminHasAllPermissions() {
        Set<Permission> adminPermissions = RolePermissions.forRole(UserRole.ADMIN);
        assertEquals(Permission.values().length, adminPermissions.size());
        for (Permission permission : Permission.values()) {
            assertTrue(adminPermissions.contains(permission), permission + " missing from ADMIN");
        }
    }

    @Test
    void employeeHasLimitedPermissions() {
        Set<Permission> employeePermissions = RolePermissions.forRole(UserRole.EMPLOYEE);

        assertTrue(employeePermissions.contains(Permission.PRODUCTS_VIEW));
        assertTrue(employeePermissions.contains(Permission.STORAGE_VIEW));
        assertTrue(employeePermissions.contains(Permission.SHIPMENTS_VIEW));
        assertTrue(employeePermissions.contains(Permission.SHIPMENTS_RECEIVE));
        assertTrue(employeePermissions.contains(Permission.SETTINGS_VIEW));
        assertTrue(employeePermissions.contains(Permission.NOTIFICATIONS_VIEW));
        // Team page is view-only for employees; admin actions inside the tabs stay gated
        // by USERS_MANAGE / role checks in the UI.
        assertTrue(employeePermissions.contains(Permission.TEAM_VIEW));

        assertFalse(employeePermissions.contains(Permission.DASHBOARD_VIEW));
        assertFalse(employeePermissions.contains(Permission.PRODUCTS_CREATE));
        assertFalse(employeePermissions.contains(Permission.ANALYTICS_VIEW));
        assertFalse(employeePermissions.contains(Permission.SHIPMENTS_CREATE));
        assertFalse(employeePermissions.contains(Permission.SHIPMENTS_UPDATE));
        assertFalse(employeePermissions.contains(Permission.STORAGE_UPDATE));
        assertFalse(employeePermissions.contains(Permission.STORAGE_DELETE));
    }

    @Test
    void assistantManagerExcludesAdminOnlyPermissions() {
        Set<Permission> assistantManagerPermissions = RolePermissions.forRole(UserRole.ASSISTANT_MANAGER);

        assertFalse(assistantManagerPermissions.contains(Permission.COSTS_VIEW));
        assertFalse(assistantManagerPermissions.contains(Permission.USERS_MANAGE));
        assertTrue(assistantManagerPermissions.contains(Permission.PRODUCTS_CREATE));
        assertTrue(assistantManagerPermissions.contains(Permission.MSRP_VIEW));
        assertTrue(assistantManagerPermissions.contains(Permission.KUJI_PRICES_VIEW));
    }

    @Test
    void onlyAdminAndAssistantManagerSeeMsrpAndKujiPrices() {
        assertFalse(RolePermissions.forRole(UserRole.EMPLOYEE).contains(Permission.MSRP_VIEW));
        assertFalse(RolePermissions.forRole(UserRole.EMPLOYEE).contains(Permission.KUJI_PRICES_VIEW));
    }

    @Test
    void onlyAdminSeesCosts() {
        assertTrue(RolePermissions.forRole(UserRole.ADMIN).contains(Permission.COSTS_VIEW));
        assertFalse(RolePermissions.forRole(UserRole.ASSISTANT_MANAGER).contains(Permission.COSTS_VIEW));
        assertFalse(RolePermissions.forRole(UserRole.EMPLOYEE).contains(Permission.COSTS_VIEW));
    }

    @Test
    void hasPermissionReturnsFalseForNullRole() {
        assertFalse(RolePermissions.hasPermission((UserRole) null, Permission.DASHBOARD_VIEW));
        assertEquals(Set.of(), RolePermissions.forRole(null));
    }

    @Test
    void hasPermissionByRoleNameIsCaseInsensitive() {
        assertTrue(RolePermissions.hasPermission("admin", Permission.COSTS_VIEW));
        assertTrue(RolePermissions.hasPermission("ADMIN", Permission.COSTS_VIEW));
    }

    @Test
    void hasPermissionByRoleNameReturnsFalseForUnrecognizedRole() {
        // JwtAuthenticationFilter's fallback for an authenticated-but-unregistered caller.
        assertFalse(RolePermissions.hasPermission("USER", Permission.PRODUCTS_VIEW));
        assertFalse(RolePermissions.hasPermission((String) null, Permission.PRODUCTS_VIEW));
    }

    @Test
    void hasPermissionByAuthenticationReturnsFalseWhenPrincipalIsNotAuthenticatedPrincipal() {
        org.springframework.security.core.Authentication authentication =
                org.mockito.Mockito.mock(org.springframework.security.core.Authentication.class);
        org.mockito.Mockito.when(authentication.getPrincipal()).thenReturn("not-a-principal");
        assertFalse(RolePermissions.hasPermission(authentication, Permission.PRODUCTS_VIEW));
        assertFalse(RolePermissions.hasPermission((org.springframework.security.core.Authentication) null, Permission.PRODUCTS_VIEW));
    }
}
